backend/routes/friends.js

```diff
@@ -88,11 +88,63 @@ router.get('/:userId', verifyToken, async (req, res) => {
         res.status(500).json({ error: 'Internal server error'})
     }
 });
+
 // friend recommendation
 router.get('/recommendedFriends/:userId', verifyToken, async (req, res) => {
     try{
         const userId = req.params.userId;
-        // fetch all post with their likes and comments
+
+          // fetch current user's friends - 1st degree
+          const friends = await Friend.findAll({ where: { userId }});
+          const friendIds = friends.map(friend => friend.friendId);
+
+          // helper function to fetch friends of  a given userId  excluding the current friends and the user
+          const excludeFriends = async (userIds, excludeIds) => {
+            return await Friend.findAll({
+                where: {
+                    userId: { [Op.in]: userIds},
+                    friendId: { [Op.notIn]: excludeIds}
+                }
+            });
+          };
+
+          // friends of friends - 2nd degree
+          const friendsOfFriends = await excludeFriends(friendIds, [...friendIds, userId]);
+
+
+          // count the occurrences of each friendId for friend of friends (2nd degree)
+          const countFriends = friendsOfFriends.reduce((acc, friend) => {
+              acc[friend.friendId] = (acc[friend.friendId] || 0) + 1;
+              return acc;
+          }, {});
+
+          // fetch friends of friends of friends 3rd degree
+          const secondDegreeFriendIds = Object.keys(countFriends).map(id => parseInt(id));
+          console.log('second:',secondDegreeFriendIds)
+          const friendsOfFriendsOfFriends = await excludeFriends(secondDegreeFriendIds, [...friendIds, userId, ...secondDegreeFriendIds]);
+
+          // count the occurences of each friendId for friend of friend of friends (3rd degree)
+          const countFriendsOfFriends = friendsOfFriendsOfFriends.reduce((acc, friend) => {
+              acc[friend.friendId] = (acc[friend.friendId] || 0) + 1;
+              return acc;
+          }, {})
+
+          //connectivity score
+          const connectivityScore = (user) => {
+            const mutualFriends = friends.filter(friend => friend.friendId === user.id).length;
+            const mutualFriendsOfFriends = friendsOfFriends.filter(friend => friend.friendId === user.id).length;
+            const mutualFriendsOfFriendsOfFriends = friendsOfFriendsOfFriends.filter(friend => friend.friendId === user.id).length;
```

Comment:
it would be great if we could get some comments explaining how these are different - examples would be lovely!

Reply:
Similarly I think this functionality can be solved with a recursive or iterative function that would actually implement the BFS. Right now we are hard-coding the BFS steps which isn't easy to scale or optimize.

Think of your user nodes and friendship edges as a graph you are trying to traverse. Finding mutuals and keeping track of when you first found them can be considered a part of the information you need to track while performing the BFS: https://www.geeksforgeeks.org/breadth-first-search-or-bfs-for-a-graph/